### PR TITLE
Init command JSON output

### LINKED ISSUE #3124: restic init command no json output
Output of `restic version`
--------------------------
`restic 0.11.0 compiled with go1.15.3 on linux/amd64`

How did you run restic exactly?
-------------------------------
`restic -vvv --json`

Output:
```
created restic repository eb2b2b23fb at /test

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Local FS

Expected behavior
-----------------
Output will be formated in json instead of plaintext.

Steps to reproduce the behavior
-------------------------------
Initialize new repository. 

Do you have any idea what may have caused this?
-----------------------------------------------
No.

Do you have an idea how to solve the issue?
-------------------------------------------
No.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
I love it, I run restic inside of my Kubernetes cluster to backup all my PVC's.
Thanks for this great tool.

### PR DESCRIPTION
**Pull request by metalsp0rk**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Enablement of JSON output on the init command.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Fixes #3124

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: cmd/restic/cmd_init.go [hunk 1]
```diff
@@ -55,7 +58,20 @@ func runInit(opts InitOptions, gopts GlobalOptions, args []string) error {
 
 	be, err := create(repo, gopts.extended)
 	if err != nil {
-		return errors.Fatalf("create repository at %s failed: %v\n", location.StripPassword(gopts.Repo), err)
+		if !gopts.JSON {
+			return errors.Fatalf("create repository at %s failed: %v\n", location.StripPassword(gopts.Repo), err)
+		} else {
+			status := struct {
+				Status     string `json:"status"`
```
Comment (MichaelEischer) [inline cmd/restic/cmd_init.go:65]: Please extract the definition into a single place.

File: cmd/restic/cmd_init.go [hunk 2]
```diff
@@ -55,7 +58,20 @@ func runInit(opts InitOptions, gopts GlobalOptions, args []string) error {
...
+				Repository string `json:"repository"`
+				Message    string `json:"message"`
+			}{
+				Status:     "error_repository",
+				Repository: location.StripPassword(gopts.Repo),
+				Message:    err.Error(),
+			}
+			return errors.Fatalf(toJSONString(status))
```
Comment (MichaelEischer) [inline cmd/restic/cmd_init.go:73]: Please use `errors.Fatal(...)` here.

File: cmd/restic/cmd_init.go [hunk 3]
```diff
@@ -69,18 +85,54 @@ func runInit(opts InitOptions, gopts GlobalOptions, args []string) error {
...
+		Verbosef("irrecoverably lost.\n")
+
+	} else {
+		status := struct {
+			Status     string `json:"status"`
+			Id         string `json:"id"`
+			Repository string `json:"repository"`
+		}{
+			Status:     "success",
+			Id:         s.Config().ID[:10],
```
Comment (MichaelEischer) [inline cmd/restic/cmd_init.go:118]: The JSON output should contain the full ID.

File: cmd/restic/cmd_init.go [hunk 4]
```diff
@@ -69,18 +85,54 @@ func runInit(opts InitOptions, gopts GlobalOptions, args []string) error {
...
+	}
 
 	return nil
 }
 
+func toJSONString(status interface{}) string {
+	buf := new(bytes.Buffer)
+	err := json.NewEncoder(buf).Encode(status)
+	if err != nil {
+		Warnf("Could not encode JSON string")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_init.go:131]: I think it's better to `panic` here. The warning would garble the (non-existent) JSON output.

File: changelog/unreleased/issue-3124 [hunk 5]
```diff
@@ -0,0 +1,6 @@
+Enhancement: Enable support for json output on init command
+
+Init command did not listen to the --json flag, it now outputs a struct depending on a successful or failed creation of a repository, with all relevant information
+
+
+https://github.com/restic/restic/issues/3124
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3124:6]: This should also link to the PR.

File: cmd/restic/cmd_init.go [hunk 6]
```diff
@@ -69,18 +85,54 @@ func runInit(opts InitOptions, gopts GlobalOptions, args []string) error {
...
+		Verbosef("created restic repository %v at %s\n", s.Config().ID[:10], location.StripPassword(gopts.Repo))
+		Verbosef("\n")
+		Verbosef("Please note that knowledge of your password is required to access\n")
+		Verbosef("the repository. Losing your password means that your data is\n")
+		Verbosef("irrecoverably lost.\n")
+
+	} else {
+		status := struct {
+			Status     string `json:"status"`
+			Id         string `json:"id"`
```
Comment (MichaelEischer) [inline cmd/restic/cmd_init.go:114]: The standard capitalization for `ID` of variables names in restic is all caps (The JSON field name is fine).

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
I've added a few minor comments to the PR. However, there's one point which requires discussion: All other places in restic just return the standard error message when an error happens and don't encode those as JSON. It might for now be better to stay consistent with the other commands.

**Comment by metalsp0rk** [comment]
>there's one point which requires discussion: All other places in restic just return the standard error message when an error happens and don't encode those as JSON. It might for now be better to stay consistent with the other commands.

So my rationale here is that it would benefit us to output errors as a JSON object because if a wrapper script is expecting JSON output, the script will not be able to easily handle non-JSON output errors.  Of course, it could just panic and alert the user, but in some cases, issues that cause failure could be automatically solved by a wrapper script.

I'm okay with staying consistent with other commands for now.  Maybe roll those error handling changes into another PR if we decide we want that, down the line.

**Comment by metalsp0rk** [comment]
I'm ready to merge this.  Can I get another review on this?

**Comment by MichaelEischer** [comment]
> I'm ready to merge this. Can I get another review on this?

I totally forgot about this PR. It's now rebased and slightly cleaned up. The `status` field is now called `message_type` as that is also used to identify the message type in other places.

The remaining part is to decide on a `message_type` name. Right now, it is `repository_created`, although something shorter like `summary` might also be suitable.

**Review by MichaelEischer — APPROVED** [review]
LGTM.

**Pull request merged**